### PR TITLE
Masq 110 key management protocol + encryption at rest implementation

### PR DESCRIPTION
**Pull request by leventdem**
Add userAppDEK alongside with userappDBid.
Encrypt/decrypt values in database.
Update MasqMock also.

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -89,20 +90,26 @@ class Masq {
...
+    window.sessionStorage.setItem('dataEncryptionKey', dataEncryptionKey)
   }
 
-  _loadSessionInfo () {
+  async _loadSessionInfo () {
     // If userId is in sesssion storage, use it and do not touch localStorage
     const sessionUserId = window.sessionStorage.getItem('userId')
-    if (sessionUserId) {
+    const sessionDataEncryptionKey = window.sessionStorage.getItem('dataEncryptionKey')
+    if (sessionUserId && sessionDataEncryptionKey) {
```
Comment (G-Ray) [inline src/index.js]: Maybe we can use only check if sessionUserId is defined. In this case the key should be stored obviously right ?

File: src/index.js [hunk 2]
```diff
@@ -113,6 +120,11 @@ class Masq {
       this.userId = localStorageUserId
       window.sessionStorage.setItem('userId', this.userId)
     }
+    const localStorageDataEncryptionKey = window.localStorage.getItem('dataEncryptionKey')
+    if (localStorageDataEncryptionKey) {
```
Comment (G-Ray) [inline src/index.js]: Same here, maybe we can directly improt the key if localStorageUserId is defined (in the `if (localStorageUserId)` block)

File: test/mockMasqApp.js [hunk 3]
```diff
@@ -36,14 +42,27 @@ class MockMasqApp {
...
+
   async get (userAppId, key) {
-    const nodes = await this.dbs[userAppId].getAsync(key)
-    if (!nodes.length) return nodes[0]
-    return nodes[0].value
+    const node = await this.dbs[userAppId].getAsync(key)
+    if (!node) return undefined
+    const dec = await this._decryptValue(node.value)
+    return dec
+    // return node.value
```
Comment (G-Ray) [inline test/mockMasqApp.js]: you can remove the comment ;)

File: test/mockMasqApp.js [hunk 4]
```diff
@@ -36,14 +42,27 @@ class MockMasqApp {
...
   }
 
   async put (userAppId, key, value) {
-    return this.dbs[userAppId].putAsync(key, value)
+    const enc = await this._encryptValue(value)
+    return this.dbs[userAppId].putAsync(key, enc)
```
Comment (G-Ray) [inline test/mockMasqApp.js]: :+1:

File: test/mockMasqApp.js [hunk 5]
```diff
@@ -84,7 +103,8 @@ class MockMasqApp {
           if (authorized) {
             const msg = {
               msg: 'authorized',
-              userAppDbId: userAppId
+              userAppDbId: userAppId,
+              userAppDEK: this.userAppDEK
```
Comment (G-Ray) [inline test/mockMasqApp.js]: ok

File: src/index.js (hunk 1 above)
Comment (sujeebant) [reply src/index.js]: :+1:

File: src/index.js (hunk 2 above)
Comment (sujeebant) [reply src/index.js]: :+1:

File: src/index.js [hunk 6]
```diff
@@ -450,8 +493,16 @@ class Masq {
...
     const db = this._getDB()
+    this._checkDEK()
     const list = await db.listAsync(prefix)
-    const reformattedDic = list.reduce((dic, e) => {
+
+    const decList = await Promise.all(list.map(async (elt) => ({
+      key: elt.key,
+      value: await this._decryptValue(elt.value)
+    })
+    ))
```
Comment (sujeebant) [inline src/index.js]: maybe put }))) on the same line ?

File: src/index.js (hunk 1 above)
Comment (leventdem) [reply src/index.js]: ok done

File: src/index.js (hunk 2 above)
Comment (leventdem) [reply src/index.js]: ok

File: test/mockMasqApp.js (hunk 3 above)
Comment (leventdem) [reply test/mockMasqApp.js]: done

**Pull request merged**